kernel/arch/i386/interrupt.S

```diff
@@ -0,0 +1,205 @@
+; Defined in isr.c
+extern isr_handler
+extern irq_handler
+align 4
+
+; Common ISR code
+isr_common_stub:
+    ; 1. Save CPU state
+    pushad              ; Pushes edi,esi,ebp,esp,ebx,edx,ecx,eax
+    mov ax, ds          ; Lower 16-bits of eax = ds.
+    push eax            ; save the data segment descriptor
+    mov ax, 0x10        ; kernel data segment descriptor
+    mov ds, ax
+    mov es, ax
+    mov fs, ax
+    mov gs, ax
+    push esp            ; Push registers_t *r
+    ; 2. Clear the direction flag (eflags) & call C handler
+    cld                 ; C code following the sysV ABI requires DF to be clear on function entry
+    call isr_handler
+
+    ; 3. Restore state
+    add esp, 4
+    pop eax
+    mov ds, ax
+    mov es, ax
+    mov fs, ax
+    mov gs, ax
+    popad
+    add esp, 8          ; Cleans up the pushed error code and pushed ISR number
+    iret                ; pops 5 things at once: CS, EIP, EFLAGS, SS, and ESP
+    ; These irets need to be iretq's when in long mode
+
+; Common IRQ code. Identical to ISR code except for the 'call'
+; and the 'pop ebx'
+irq_common_stub:
+    pushad
+    mov ax, ds
+    push eax
+    mov ax, 0x10
+    mov ds, ax
+    mov es, ax
+    mov fs, ax
+    mov gs, ax
+    push esp
+    cld
+    call irq_handler ; Different than the ISR code
+    add esp, 4
+    pop ebx          ; Different than the ISR code
+    mov ds, bx
+    mov es, bx
+    mov fs, bx
+    mov gs, bx
+    popad
+    add esp, 8
+    iret
+    ; These irets need to be iretq's when in long mode
+
+; Macro expansions for isr and irq
+
+%macro m_isr_2 1
+
+global isr%1
+isr%1:
+    push 0
+    push %1
+    jmp isr_common_stub
+
+%endmacro
+
+%macro m_isr_1 1
+
+global isr%1
+isr%1:
+    push %1
+    jmp isr_common_stub
+
+%endmacro
+
+%macro m_irq_2 2
+
+global irq%1
+irq%1:
+    push %1
+    push %2
+    jmp irq_common_stub
+
+%endmacro
+
+; We don't get information about which interrupt was called
+; when the handler is run, so we will need to have a different handler
+; for every interrupt.
+; Furthermore, some interrupts push an error code onto the stack but others
+; don't, so we will push a dummy error code for those which don't, so that
+; we have a consistent stack for all of them.
+
+; All of the following Interrupt Service Routines are
+; exception handlers. As such, they push the error code
+; along with their exception (interrupt) number. If an
+; exception has an error code, $0 will not be pushed as
+; the CPU already does that for us.
+;
+m_isr_2 0 ; 0: Divide By Zero Exception
+
```

Comment:
I would remove the whitespace in between the ISRs and IRQs. It'll be nice to have more condensed source now that we have macros.

Reply:
No worries.  Just want to make sure I understand what you are asking for.  You'd like to see this:

m_isr_2 0 ; 0: Divide By Zero Exception
m_isr_2 1 ; 1: Debug Exception
m_isr_2 2 ; 2: Non Maskable Interrupt Exception
...

Reply:
Removed whitespace to make isr / irq as above.